java/src/main/java/io/cucumber/java/JavaHookDefinition.java

```diff
@@ -22,41 +32,108 @@
-                .build();
+    private static InvalidMethodSignatureException createInvalidSignatureException(
+            Method method,
+            Class<?>[] methodAnnotations,
+            Set<Class<?>> acceptedTypes
+    ) {
+        InvalidMethodSignatureException.InvalidMethodSignatureExceptionBuilder builder = builder(method);
+        if (methodAnnotations.length == 0) {
+            throw new IllegalArgumentException(
+                "Method should be annotated with one of: @Before, @After, @BeforeStep, @AfterStep");
+        }
+        builder.addAnnotations(methodAnnotations);
+        String methodName = "public void before_or_after";
+        if (acceptedTypes.contains(Step.class)) {
+            methodName += "_step";
+            builder.addSignature(methodName + "(io.cucumber.java.Scenario scenario, io.cucumber.java.Step step)");
+            builder.addSignature(methodName + "(io.cucumber.java.Step step, io.cucumber.java.Scenario scenario)");
+            builder.addSignature(methodName + "(io.cucumber.java.Step step)");
+        }
+        builder.addSignature(methodName + "(io.cucumber.java.Scenario scenario)");
+        builder.addSignature(methodName + "()");
+        return builder.build();
     }
 
     @Override
     public void execute(TestCaseState state) {
-        Object[] args;
-        if (method.getParameterTypes().length == 1) {
-            args = new Object[] { new io.cucumber.java.Scenario(state) };
-        } else {
-            args = new Object[0];
+        List<Object> parameters = new ArrayList<>();
+        for (Class<?> parameterType : method.getParameterTypes()) {
+            if (parameterType.equals(Scenario.class)) {
+                parameters.add(new Scenario(state));
+            }
+            if (parameterType.equals(Step.class)) {
+                PickleStep pickleStep = getCurrentPickleStep(state);
+                parameters.add(new Step(pickleStep));
+            }
         }
+        invokeMethod(parameters.toArray());
+    }
 
-        invokeMethod(args);
+    private PickleStep getCurrentPickleStep(TestCaseState state) {
+        Optional<io.cucumber.core.backend.Step> currentTestStep = state.getCurrentTestStep();
+        if (!currentTestStep.isPresent()) {
+            throw new IllegalStateException("No current TestStep was found in TestCaseState");
+        }
+        if (currentTestStep.get() instanceof HookStep) {
+            PickleStep currentStep = ((HookStep)currentTestStep.get()).getRelatedStep();
+            if (null == currentStep) {
+                throw new IllegalStateException("Current HookStep has no related PickleStep");
+            }
+            return currentStep;
+        }
+        throw new IllegalArgumentException(
+                String.format("Current TestStep should be a %s instead of a %s",
+                        HookStep.class.getSimpleName(), currentTestStep.get().getClass().getSimpleName()));
```

Comment:
```suggestion
        io.cucumber.core.backend.Step step = state.getCurrentTestStep()
                .orElseThrow(() -> new IllegalStateException("No current TestStep was found in TestCaseState"));

        HookStep hookStep = Optional.of(step)
                .filter(HookStep.class::isInstance)
                .orElseThrow(() -> new IllegalArgumentException(String.format("Current TestStep should be a %s instead of a %s",
                        HookStep.class.getSimpleName(), currentTestStep.get().getClass().getSimpleName())));
                                
        return Optional.of(hookStep)
                .map(HookStep::getRelatedStep)
                .orElseThrow(() -> new IllegalStateException("Current HookStep has no related PickleStep"));

```